### DISCUSSION TITLE
Help me for analysis logs for relove problem my clients disconnect and reconnent

### DISCUSSION THREAD

**Discussion by reekoong** [emqx/emqx#4699] [Q&A]
Hello, I would like your help me for check the MQTT error of my system.
I use esp32 with pubsubclient
1. My client connet broker with unique id
2. My client public LDR value( ex. 2000 ,2000,2000,3000,4000) every 2 sec.
3. I use default keepalive with my library (15 sec)
4. My EMQ install on google cloud compute engine with centos7
Then, I have problem my mqtt broker often offline, Client disconnect and reconnect. and sometime my broker will disconnect a client when i publish message to the client has subscribe.

Thank

my logs
`arse failed for function_clause
[{emqx_frame,parse_utf8_string,[<<1,0,198>>],[{file,"emqx_frame.erl"},{line,424}]},{emqx_frame,parse_packet,3,[{file,"emqx_frame.erl"},{line,170}]},{emqx_frame,parse_frame,4,[{file,"emqx_frame.erl"},{line,147}]},{emqx_connection,parse_incoming,3,[{file,"emqx_connection.erl"},{line,519}]},{emqx_connection,handle_msg,2,[{file,"emqx_connection.erl"},{line,512}]},{emqx_connection,process_msg,3,[{file,"emqx_connection.erl"},{line,293}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]`

`Parse failed for function_clause
[{emqx_frame,parse_packet,[{mqtt_packet_header,0,false,0,true},<<0,24,100,98,111,116,48,48,48,49,47,109,111,110,105,116,111,114,47,115,116,97,116,117,115,47,99,111,111,108,105,110,103>>,#{max_size => 1048576,strict_mode => false,version => 4}],[{file,"emqx_frame.erl"},{line,169}]},{emqx_frame,parse_frame,4,[{file,"emqx_frame.erl"},{line,147}]},{emqx_connection,parse_incoming,3,[{file,"emqx_connection.erl"},{line,519}]},{emqx_connection,handle_msg,2,[{file,"emqx_connection.erl"},{line,512}]},{emqx_connection,process_msg,3,[{file,"emqx_connection.erl"},{line,293}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]`

**Comment by k32** [comment]
Hello,

Which version of EMQ X broker are you using? Could you clarify more specifically the version of the client? 
How often do you see these errors? Would it be possible to collect tcpdump?

**Reply by reekoong** [reply]
I use emqx 4.2.9 , I have this problem every 1 hour from every client (I have 12 client runing now). I can allow you to access to my server.

**Reply by k32** [reply]
You can write me a mail to `redacted`

**Reply by reekoong** [reply]
Ok.

**Reply by k32** [reply]
Thanks, I collected the emqx logs, I think it should be sufficient to start investigation.

**Reply by k32** [reply]
I analyzed the crash report from MQTT data, and it looks like the client sent a malformed MQTT Connect packet with type reserved. If I am not mistaken, it's prohibited by MQTT standard. I will try to run tcpdump on the live traffic, to make sure it's the exact data received from the client, and not some sort of parser bug.

**Reply by reekoong** [reply]
I’m ready to give you more information if you need it. For you to analyze better.Thank

**Reply by k32** [reply]
Hi!
Sorry for the delay. I will continue investigating the issue soon. Some findings so far: a lot of crashes in the emqx_frame parser were caused by non-MQTT traffic hitting your endpoint. For example:
```
2021-04-22 23:18:45.868 [error] 23.129.64.251:44043 [MQTT] 
Parse failed for function_clause
....................
Frame data:<<"GET / HTTP/1.1\r\nHost: 34.87.52.254:1883\r\nUser-Agent: Mozilla/5.0 (compatible; Baiduspider/2.0; +http://www.baidu.com/search/spider.html)\r\nConnection: close\r\nAccept: */*\r\nAccept-Encoding: gzip\r\n\r\n">>
```

For some reason Baidu spider assumed that your endpoint is HTTP and decided to scrape it. I am not sure it's related to your clients disconnecting (likely not), so I will continue looking into it.

**Reply by reekoong** [reply]
I think so too. Now i think, Did data of a client as send public it's will have problem?
Would there be any possibility of a client send public to many topic (5 topic) in near time? 
e.g.
pub ---> topic1
delay 100ms
pub ---> topic2
delay 100ms
pub ---> topic3
delay 100ms
pub ---> topic4
delay 100ms
pub ---> topic5
delay 100ms

**Reply by zmstone** [reply]
it should be fine for clients to publish to multiple topics

**Reply by reekoong** [reply]
@k32 I'm just change log level to debug.
@zmstone Thank for commented.

**Comment by zmstone** [comment]
just one hint: `proc_lib,wake_up` implies that the process went into hibernation before.

**Reply by zmstone** [reply]
the default hibernation threshold is 1 minute,
you may want to check if the 15-seconds keep-alive was actually working.

**Reply by reekoong** [reply]
Actually a client send public of sensor value every 2 sec. I try use keepalive to 60 sec but not resolve. Sometime mqtt broker will disconnect at once when send public to topic at client subscribe.

**Reply by zmstone** [reply]
actually setting keepalive to 60 only increases the chance of process going into hibernation.
if your publisher sends data every 2 seconds, there is no need for keepalive.

could you check if the crash log having `wake_up` in it are from publishers ?
there should be client ID, username and peer IP in the log.

**Reply by reekoong** [reply]
Can you guide me to find a wake_up , proc_lib file path?

**Reply by zmstone** [reply]
it's in your logs:

Parse failed for function_clause...... [{file,"emqx_connection.erl"},{line,293}]},{**proc_lib,wake_up**,3,[{file,"proc_lib.erl"},{line,259}]}]

before "Parse failed" there should be peer information, clientid and/or username.
please check if this is a client which is publishing every 2 seconds.

Asking because (by looking at the logs you provided) we have very little knowledge about which kind of incoming message caused the process to crash

**Reply by reekoong** [reply]
Okay, I'm got it. Can i send log file to you if you need. Just told you email to me.
Thank you.

**Reply by zmstone** [reply]
In order to help narrow down the investigation scope, please
1. change to `log.level = debug` in `emqx.conf` and restart the node
2. start as less client as possible to reproduce the crash and share emqx logs.

**Reply by zmstone** [reply]
the logs can be shared here in github discussion.

**Reply by reekoong** [reply]
I'm just to change log.level to debug, I will share the logs in soon.

**Reply by reekoong** [reply]
[logs.tar.gz](https://github.com/emqx/emqx/files/6411716/logs.tar.gz)

**Reply by zmstone** [reply]
Hi @reekoong could you give an example clientid which get disconnected by the broker ?

**Reply by reekoong** [reply]
Ok,Actually all clientID has problems ,But you can use client id wbot0003,dbot0003,wbot0007 for example.

**Reply by zmstone** [reply]
I can see there were a few crashes before 2021-05-02 21:50:29 (before you enabled debug logging).
but after that, I can't seem to find any error logs.

**Comment by zmstone** [comment]
for wbot0003, the crash was due to `0` (reserved) in the fix packet header.
0 is forbidden per spec, see: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718020

<0,30,119,98,111,116,48,48,48,51,47,109,111,110,105,116,111,114,47,108,100,114,47,114,111,116,97,114,121,76,69,68,48>>
this stream is parsed by EMQ X as 
0 -> reserved, forbidden.
then the rests of this frame is not parsed at all.


Though the remaining bytes look like a publish request.
30 -> the length of a topic
119.....68 -> the topic name: `wbot0003/monitor/ldr/rotaryLED`
the last byte 48 -> a single char `'0'`, the payload.

I'd question why the client uses a forbidden control packet for a publish request.

**Reply by zmstone** [reply]
And later after the debug log is enabled, we can see the same client, managed to publish this message successfully:

```
2021-05-02 21:50:31.497 [info] <<"wbot0003">>@49.228.162.157:52728 PUBLISH to wbot0003/monitor/ldr/rotaryLED: <<"0">>
```

**Reply by zmstone** [reply]
RAW bytes for a successful publish:
`<<48,33,0,30,119,98,111,116,48,48,48,51,47,109,111,110,105,116,111,114,47,108,100,114,47,114,111,116,97,114,121,76,69,68,48>>`
so I was wrong about `30` being the length, `30` is actually the LSB of topic-lenght, `0` is the MSB.

it all starts making sense if we try to parse `48`, `33`:
`48` = `0011,0000` = publish, no duplication, QoS0, no-retain.
`33` = the remaining bytes in this packet.
`0, 30` = the length of the topic
`119.....68` = the topic name: `wbot0003/monitor/ldr/rotaryLED`
`48` = payload `"0"`.

**Reply by zmstone** [reply]
@reekoong you may need to check in the producer code why it sometimes have the control packet + remaining length missing in the stream.

**Reply by reekoong** [reply]
Thank you for analysis.
I don't know why a client uses a forbidden control packet. Would it's be possibility  often a client it will send public payload  = 0 to broker? 

And my brocker running on google cloud engine with centos7. Does it have anything to do with google cloud firewall?

**Reply by zmstone** [reply]
It's more likely that the client 'thought' that it has already sent the control header,
hence it sent the remaining bytes directly, then the broker took the first two bytes
of remaining stream as control packet.

My experience with ESP32 is quite limited.
It looks more clear now, that it's a bug on the client side.
I'd suggest to wait for the same issue to happen again while you have the debug log enabled.
so we know which bytes the client has sent before the exception observed on the broker side.

WRT to payload:
I would not say payload `"0"` is abnormal, as it seemingly to be a 'switched off' signal published by the LED?

**Reply by reekoong** [reply]
I have a the new logs of client disconnect . You try find in clientid: wbot0006 since 2016-05-03 16.00.00

**Reply by zmstone** [reply]
Hi @reekoong 
do you mean this?

```
2021-05-03 16:10:20.692 [info] <<"wbot0006">>@49.228.162.157:50777 PUBLISH to wbot0006/monitor/status/: <<"offline">>
2021-05-03 16:10:20.692 [debug] <<"wbot0006">>@49.228.162.157:50777 [MQTT] Terminated due to {shutdown,tcp_closed}
``` 

This is due to client side closed the TCP socket.
and the "offline" message indicates a normal shutdown.

**Reply by reekoong** [reply]
offline in this case is LWT, But i don't know why it's disconnect.

**Reply by reekoong** [reply]
My problem is broker often disconnect client.

**Reply by zmstone** [reply]
From the logs you shared:
1. before debug was enabled, the disconnect was due to malformed request
2. after debug was enabled, the disconnect we saw was done by the client

please keep the debug logs open, maybe we need to wait longer for it to happen again

**Reply by zmstone** [reply]
but you shared this before ?

**Reply by zmstone** [reply]
and which client had connection closed by broker in your shared latest log ?

**Reply by reekoong** [reply]
wbot0009 wbot0005 wbot0001 dbot0003 dbot0002 wbot0006

**Reply by zmstone** [reply]
I can not find anything unusual for `wbot0001`, could you help to narrow down to when abnormal disconnect happened?

By the way, I found below 4 `tcp_closed` events, but not any DISCONNECT packets.
given that the clients sent a "offline" before `tcp_closed`, I'd assume this is how your client behaves.
```
emqx.log.1:7641:104:2021-05-04 15:00:35.453 [debug] <<"wbot0009">>@49.228.162.157:52061 [MQTT] Terminated due to {shutdown,tcp_closed}
emqx.log.3:56893:104:2021-05-04 14:24:59.682 [debug] <<"wbot0006">>@49.228.162.157:61308 [MQTT] Terminated due to {shutdown,tcp_closed}
emqx.log.5:14881:104:2021-05-04 14:46:08.735 [debug] <<"wbot0003">>@49.228.162.157:50597 [MQTT] Terminated due to {shutdown,tcp_closed}
emqx.log.5:26255:104:2021-05-04 14:49:19.459 [debug] <<"MqttClient">>@64.62.197.92:48138 [MQTT] Terminated due to {shutdown,tcp_closed}
```

**Reply by reekoong** [reply]
wbot0009 client disconnent last on 2021-05-04 15:00:35

**Reply by zmstone** [reply]
from the logs, it seems pretty clear to me :
`wbot0009` itself first posted an `"offline"` status, then closed the socket.
i.e. not the broker who closed the connection.

```
2021-05-04 15:00:35.453 [info] <<"wbot0009">>@49.228.162.157:52061 PUBLISH to wbot0009/monitor/status/: <<"offline">>
2021-05-04 15:00:35.453 [debug] <<"mqttjs_bc8d9eda">>@34.87.54.77:50600 [MQTT] SEND PUBLISH(Q0, R0, D0, Topic=wbot0009/monitor/status/, PacketId=undefined, Payload=<<"offline">>)
2021-05-04 15:00:35.453 [debug] <<"wbot0009">>@49.228.162.157:52061 [MQTT] Terminated due to {shutdown,tcp_closed}
```

**Reply by reekoong** [reply]
> given that the clients sent a "offline" before tcp_closed, I'd assume this is how your client behaves.

Sometimes I often experience disconnection when i send payload to topic at client subscribed.

e.g.
client --> sub topic "wbot0009/command"
I'm ---> pub topic "wbot0009/command" , payload "m1" or "20"
broker --> disconect for client wbot0009
client ---> pub LWT "offline"
client --> reconnent

I ever have error on mosquito mqtt logs is : 
`[info] <0.25824.7>MQTT connection "xx.xx.xx.xx:56547 -> xx.xx.xx.xx:1883" will terminate because peer closed TCP connection
[info] <0.26305.7> Closing all channels from connection 'xx.xx.xx.xx:56547 -> 10.148.0.8:1883' because it has been closed`

**Reply by reekoong** [reply]
> from the logs, it seems pretty clear to me :
> `wbot0009` itself first posted an `"offline"` status, then closed the socket.
> i.e. not the broker who closed the connection.
> 
> ```
> 2021-05-04 15:00:35.453 [info] <<"wbot0009">>@49.228.162.157:52061 PUBLISH to wbot0009/monitor/status/: <<"offline">>
> 2021-05-04 15:00:35.453 [debug] <<"mqttjs_bc8d9eda">>@34.87.54.77:50600 [MQTT] SEND PUBLISH(Q0, R0, D0, Topic=wbot0009/monitor/status/, PacketId=undefined, Payload=<<"offline">>)
> 2021-05-04 15:00:35.453 [debug] <<"wbot0009">>@49.228.162.157:52061 [MQTT] Terminated due to {shutdown,tcp_closed}
> ```

Do you know, How i will should resolved this problem?

**Reply by zmstone** [reply]
> Sometimes I often experience disconnection when i send payload to topic at client subscribed.

I'd need debug logs to understand better what happened in such cases.
In the logs you shared so far, it does not look like the broker initiated any disconnect.

**Reply by zmstone** [reply]
> > from the logs, it seems pretty clear to me :
> > `wbot0009` itself first posted an `"offline"` status, then closed the socket.
> > i.e. not the broker who closed the connection.
> > ```
> > 2021-05-04 15:00:35.453 [info] <<"wbot0009">>@49.228.162.157:52061 PUBLISH to wbot0009/monitor/status/: <<"offline">>
> > 2021-05-04 15:00:35.453 [debug] <<"mqttjs_bc8d9eda">>@34.87.54.77:50600 [MQTT] SEND PUBLISH(Q0, R0, D0, Topic=wbot0009/monitor/status/, PacketId=undefined, Payload=<<"offline">>)
> > 2021-05-04 15:00:35.453 [debug] <<"wbot0009">>@49.228.162.157:52061 [MQTT] Terminated due to {shutdown,tcp_closed}
> > ```
> 
> Do you know, How i will should resolved this problem?

I'm afraid I do not know much about how your client is implemented.

**Reply by zmstone** [reply]
> I use espwroom-32 with pubsubsclient library. Can i send example c++ code (Aduino file) to you on email for check?

Sorry. I'm afraid I can't help you there. My aduino & C++ knowledge is near to zero

**Comment by Silvenga** [comment]
@reekoong did you figure anything out? Also using a ESP32.

Unfortunately my mqtt knowledge is next to zero. 😹
